src/tablenavigation.js

```diff
@@ -0,0 +1,515 @@
+		const mapper = this.editor.editing.mapper;
+
+		const startPosition = direction == 'forward' ? boundaries.start : boundaries.end;
+
+		const treeWalker = new TreeWalker( { direction, boundaries, startPosition } );
+
+		for ( const { nextPosition, item } of treeWalker ) {
+			if ( schema.checkChild( nextPosition, '$text' ) ) {
+				const viewElement = mapper.toViewElement( item );
+
+				if ( viewElement && !viewElement.hasClass( 'ck-hidden' ) ) {
+					return nextPosition;
+				}
+			}
+		}
+	}
+
+	/**
+	 * Checks if the `modelRange` renders to a single line in the DOM.
+	 *
+	 * @private
+	 * @param {module:engine/model/range~Range} modelRange Current table cell content range.
+	 * @param {Boolean} isForward The expected navigation direction.
+	 * @returns {Boolean}
+	 */
+	_isSingleLineRange( modelRange, isForward ) {
+		const model = this.editor.model;
+		const editing = this.editor.editing;
+		const domConverter = editing.view.domConverter;
+
+		// Wrapped lines contain exactly the same position at the end of current line
+		// and at the beginning of next line. That position's client rect is at the end
+		// of current line. In case of caret at first position of the last line that 'dual'
+		// position would be detected as it's not the last line.
+		if ( isForward ) {
+			const probe = model.createSelection( modelRange.start );
+
+			model.modifySelection( probe );
+
+			// If the new position is at the end of the container then we can't use this position
+			// because it would provide incorrect result for eg caption of image and selection
+			// just before end of it. Also in this case there is no "dual" position.
+			if ( !probe.focus.isAtEnd && !modelRange.start.isEqual( probe.focus ) ) {
+				modelRange = new ModelRange( probe.focus, modelRange.end );
+			}
+		}
+
+		const viewRange = editing.mapper.toViewRange( modelRange );
+		const domRange = domConverter.viewRangeToDom( viewRange );
+		const rects = Rect.getDomRangeRects( domRange );
+
+		let boundaryVerticalPosition;
+
+		for ( const rect of rects ) {
+			if ( boundaryVerticalPosition === undefined ) {
+				boundaryVerticalPosition = Math.round( rect.bottom );
+				continue;
+			}
+
+			// Let's check if this rect is in new line.
+			if ( Math.round( rect.top ) >= boundaryVerticalPosition ) {
+				return false;
+			}
+
+			boundaryVerticalPosition = Math.max( boundaryVerticalPosition, Math.round( rect.bottom ) );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Moves the selection from the given table cell in the specified direction.
+	 *
+	 * @private
+	 * @param {module:engine/model/element~Element} tableCell The table cell to start the selection navigation.
+	 * @param {'left'|'up'|'right'|'down'} direction Direction in which selection should move.
+	 */
+	_navigateFromCellInDirection( tableCell, direction ) {
+		const model = this.editor.model;
+
+		const table = findAncestor( 'table', tableCell );
+		const tableMap = [ ...new TableWalker( table, { includeSpanned: true } ) ];
```

Comment:
Calculating this and other stuff in tables is prone to errors. Most likely when modifying a table but even those calculations could be unintentionally be wrong.

In other words, please use `TableUtils` for those:

1.  `const lastRow = tableUtils.getRows( table ) - 1`
2.  `const lastColumn = tableUtils.getColumns( table ) - 1`
3.  `const { row, column } = tableUtils.getCellLocation( tableCell )`

I'd feel safer to have this using `TableUtils`. Also using the above we don't have to iterate over spanned cells.

Reply:
As @niegowski mentionend `table` map is reused also to navigate so it makes sense to leave it as-is.